jar/jar.go

```diff
@@ -405,6 +414,22 @@ func (c *checker) checkFile(zf *zip.File, depth int, size int64, jar string) err
 	return nil
 }
 
+func readFull(r io.Reader, fi os.FileInfo, buf []byte) ([]byte, error) {
+	if !fi.Mode().IsRegular() {
+		return io.ReadAll(r) // If not a regular file, size may not be accurate.
+	}
+	if size := int(fi.Size()); cap(buf) < size {
+		buf = make([]byte, size)
```

Comment:
nit: it's kinda weird that this would allocate despite being passed a buffer. Maybe something like

```go
// readAll is a io.ReadAll alternative that optionally uses a buffer from this program's global sync.Pool
// for sufficiently small files. This increases performance when unpacking small JARs within JARs.
func readAll(r io.Reader, fs os.FileInfo) (b []byte, cleanup func(), err error) {
	size := int(fi.Size())
	if !fi.Mode().IsRegular() || size > bufSize {
		b, err = io.ReadAll(r) // If not a regular file, size may not be accurate.
        return b, func() {}, err
	}

    buf := bufPool.Get().([]byte)[:size]
    if _, err := io.ReadFull(r, buf); err != nil {
		bufPool.Put(buf)
		return nil, nil, err
    }
	return buf, func() { bufPool.Put(buf) }, nil
}
```

Then defer the cleanup func?

```go
buf, cleanup, err := readAll(f, fi)
if err != nil {
	// ...
}
defer cleanup() // Free buffer.
```

Reply:
IMHO, it's not that weird. The same behaviour can be observed in `append()` and at least some other APIs I've seen but can't easily recall.

```
// readAll is a io.ReadAll alternative that optionally uses a buffer from this program's global sync.Pool
// for sufficiently small files. This increases performance when unpacking small JARs within JARs.
```

I don't see the implementation you posted have a size limit. What should it be, if any?

Reply:
Oh, I see. You're calling Put() on the byte slice that's returned. Nvm, please see my other comment